Render world locations

jkempster34 opened the pull request:
### Add locales for world locations
This adds the necessary locales to render the `/world` page.

### Add world location index to the README
The README contains a list of live examples. Now that the world index page is
soon to be rendered by Collections, we can add it in.

### Make organisation list filtering JavaScript more generic
Before this JavaScript was moved from Whitehall to Collections, it was shared
between the `/world` page and the `/government/organisations` page. However,
since being moved it has been made specific to the `/government/organisations`
page.

Now that we are also moving the `/world` page to Collections, we want to make
this module more generic so that it can be shared again.

This also removes `data-filter-list` as it is unused.

### Render a basic world index page
Render a basic page for `/world` including the title and additional links.

### Render world locations on the world index page
The world index page displays a list of world locations sorted by letter. These
are already sorted alphabetically in the content item, so we only need to group
them.

### Render international delegations on the world index page
The world index page displays a list of world locations sorted by letter. These
are already sorted alphabetically in the content item.

### Add JavaScript filter to world index
The world index page contains a JavaScript based search. We want to use the
`list-filter` module for this.

As the world locations on the page are grouped into letters, we need to add the
concept of an `inner-block` that can be hidden when no results are present.


---

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

brucebolt commented on spec/presenters/world_index_presenter_spec.rb:
```diff
@@ -0,0 +1,14 @@
+RSpec.describe WorldIndexPresenter do
+  let(:fixture) { fetch_fixture("world_index") }
```
Are we able to load the example of content schemas in Publishing API?  That would save having such a large JSON file here.

jkempster34 replied:
Done that, thank you!

brucebolt reviewed (commented):
Looks good to me (from a backend perspective).  Just a small question on whether we can remove the JSON fixture file.

I'd suggest getting a frontend dev in the owning team to look at the JavaScript and CSS too.

hannako commented on app/views/world/index.html.erb:
```diff
@@ -46,6 +46,37 @@
+      <header class="type-heading">
+        <h2 class="world-locations-heading"><%= I18n.t("world_locations.type.international_delegation", count: 2) %></h2>
+        <p class="govuk-visually-hidden"><%= t('world_locations.count_prefix') %>
+          <span class="js-accessible-item-count"><%= @presented_index.international_delegations_count %></span> <%= I18n.t("world_locations.type.international_delegation", count: 2) %>
```
Style comment

This is just a style thing, so feel free to ignore. But if we just have `@presented_index.international_delegation.count` here instead, we could lose `delegate :count, to: :international_delegations, prefix: true` from the presenter. Likewise for `world_locations_count`. Two lines of code replaced by two `.`'s :-) ... I think??

jkempster34 replied:
I'm going to leave this because we have have to make `world_locations` public and I like that it's private because its more clear to any future developers that its not used in the view

hannako reviewed (commented):
Thanks for taking the time to give such a great PR description. I've made a few comments on the Ruby code.

jkempster34 commented on app/controllers/world_controller.rb:
```diff
@@ -0,0 +1,7 @@
+class WorldController < ApplicationController
+  def index
+    index = WorldIndex.find!("/world")
+    @presented_index = WorldIndexPresenter.new(index)
```
I think this will only get loaded once per request so there isn't any benefit to memoizing it

jkempster34 commented on app/presenters/world_index_presenter.rb:
```diff
@@ -0,0 +1,9 @@
+    @world_index = world_index
+  end
+
+  def title
```
We have the instance variable `@content_item` as a Hash (from `setup_content_item_and_navigation_helpers`), but that means we would have to access the title in the view as `@content_item["title"]` which I don't think is as nice.

I think if we alternatively get rid of this method, and call `@world_index.content_item.title` in the view, we are ignoring the Law of Demeter (not sure how important that is tbh 🤷) and going against the [other current implementations like this one ](https://github.com/alphagov/collections/blob/59183090590376a8e101ae609b2ae5e72065f3be/app/presenters/organisations/index_presenter.rb#L30)

What do you think? I may have misunderstood you entirely

jkempster34 commented on app/assets/javascripts/application.js:
```diff
@@ -9,6 +9,6 @@
 //= require govuk_publishing_components/components/step-by-step-nav
 
 //= require support
-//= require modules/organisation-list-filter
```
Done, the `filter_terms` are added as attributes on the list items so the JS can filter them without searching for specific classes 
<img width="515" alt="image" src="https://github.com/alphagov/collections/assets/47089130/6e8197aa-1f79-4dd5-bd2e-e0bceffdb3b1">

jkempster34 commented on app/presenters/world_index_presenter.rb:
```diff
@@ -1,9 +1,42 @@
+      .sort
+  end
+
+  def world_location_link(world_location)
```
Not sure what to do about this, I quite like it as is because it's easily testable. What do you mean about the tracking, and how strongly do you feel about it?

hannako replied:
We can refactor it if/when tracking is required. Fine to leave for now 👍

jkempster34 commented:
> Thanks for taking the time to give such a great PR description. I've made a few comments on the Ruby code.

Thanks for the thorough review 🙌  I've addressed most points, but left a few questions as well

hannako reviewed (approved):
Looks great to me! I know @hannalaakso was wanting to review the javascript

patrickpatrickpatrick commented:
Javascript looks good to me, very nice consolidation of the search terms in the presenter meaning we can generalise the filter script.
